### ISSUE TITLE
model and repository of clean achitecture for creating game

### ISSUE THREAD

**Issue by metalalive** [Game-as-a-Service/Pandemic-Reign-of-Cthulhu#11]
The issue #10 has been completed.

According to the functions declared in the repository and domain models , we need someone to implement the detail . 
- Domain models
  - `Game`
    - `add_players()` , ensure total number of players won't exceed the limit (at most 4 people)
    - `id` , identifier could be generated in UUID format ?
    - `assign_character()`, ensure each player selected different investigator, hash table can do this well, or any other option ?
  - `Player`
    - record the investigator the player has selected
    - is it good idea to bind `Player` instances to specific `Game` instance ?
- In-memory repository
  - `InMemoryRepository.save()` , use hash table to store key (game ID) value (relevant `Game` instance) pairs ?

**Comment by teds-lin** [comment]
> * Domain models
>   * `Game`
>     * `id` , identifier could be generated in UUID format ?

1.  in-memory情境我會建議用UUID，具體依照用途，UUID有不同版本可供選擇，可參考此[文件](https://docs.rs/uuid/latest/uuid/#which-uuid-version-should-i-use)
**TLDR**:
目的單純僅需要隨機選v4 (執行效率考量)
需要用作資料庫索引或時間排序用途選v7
2.  之後實作用db的話，我建議直接用db的id()
>    *  `assign_character()`, ensure each player selected different investigator, hash table can do this well, or any other option ?

我認為hash table是好選擇

>   * `Player`
>     * is it good idea to bind `Player` instances to specific `Game` instance ?

我認為ok, 這隱含假設一位玩家同一時間只能進行一場遊戲

> * In-memory repository
>   
>   * `InMemoryRepository.save()` , use hash table to store key (game ID) value (relevant `Game` instance) pairs ?

感覺ok, 沒有其他更好的想法

**Comment by metalalive** [comment]
> 目的單純僅需要隨機選v4 (執行效率考量)
需要用作資料庫索引或時間排序用途選v7
在我們case v4足夠使用了

我之前讀過 [RFC 4122](https://datatracker.ietf.org/doc/html/rfc4122) 文件 + [相關的 draft released this year](https://datatracker.ietf.org/doc/draft-ietf-uuidrev-rfc4122bis/), v4 ,v7 兩者都OK, 
先用v4也可以

> 之後實作用db的話，我建議直接用db的id()

之後我將在 [GitHub discussion](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/discussions) 中開一個討論, 看要使用哪個資料庫 (e.g. MongoDB, PostgreSQL, any other Non-relational database)


Teds 你能處理這個Issue嗎 ?
目前我正在處理另一個 Issue #19 , 修改完 open API 規格後,
我將開始寫 code 加入一些 API endpoints 供玩家在遊戲開始前進行設定
我的想法是, backend dev 可以同時工作 / 完成一些事情
這你ok嗎 ? 謝謝

**Comment by teds-lin** [comment]
> Teds 你能處理這個Issue嗎 ?

ok, 不過今天比較忙，我明天處理
